VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,545 @@
+     */
+    function getDatacenters(uint[] calldata ids) external view override returns(Datacenter[] memory) {
+        uint count = ids.length;
+        Datacenter[] memory list = new Datacenter[](count);
+
+        for (uint i; i < count; ++i) {
+            if (!_datacenterInfo[ids[i]].isActive) {
+                revert InvalidDatacenterId(ids[i]);
+            }
+
+            list[i] = _dataCenterMap[ids[i]];
+        }
+
+        return list;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getDataCentersByCountry(string calldata countryCode) external view override returns(Datacenter[] memory) {
+        validateCountryCode(countryCode);
+        
+        uint count = _countryDataCenterIds[countryCode].length();
+        Datacenter[] memory list = new Datacenter[](count);
+
+        EnumerableSetUpgradeable.UintSet storage countryIds = _countryDataCenterIds[countryCode];
+
+        for (uint i; i < count; ++i) {
+            list[i] = _dataCenterMap[countryIds.at(i)];
+        }
+
+        return list;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getDataCentersByRegion(string calldata regionCode) external view override returns(Datacenter[] memory) {
+        validateRegionCode(regionCode);
+        
+        uint count = _regionDataCenterIds[regionCode].length();
+        Datacenter[] memory list = new Datacenter[](count);
+
+        EnumerableSetUpgradeable.UintSet storage regionIds = _regionDataCenterIds[regionCode];
+
+        for (uint i; i < count; ++i) {
+            list[i] = _dataCenterMap[regionIds.at(i)];
+        }
+
+        return list;
+    }
+
+    /**
+     * @notice Check the `authSignature` parameter of `addNode()` function
+     * @param didAddress DID address that is associated with the storage node
+     * @param authSignature Signature signed by a trusted signer
+     */
+    function verifyAuthSignature(address didAddress, bytes calldata authSignature) internal view {
+        EnumerableSetUpgradeable.AddressSet storage signers = _trustedSigners;
+
+        if (signers.length() == 0) {
+            revert NoSigners();
+        }
+
+        if (authSignature.length == 0) {
+            revert InvalidSignature();
+        }
+
+        bytes memory rawMsg = abi.encodePacked(didAddress);
+        bytes32 msgHash = keccak256(rawMsg);
+
+        address authSigner = ECDSAUpgradeable.recover(msgHash, authSignature);
+
+        bool isVerified;
+        uint index;
+        
+        while (index < signers.length() && !isVerified) {
+            address account = signers.at(index);
+
+            if (authSigner == account) {
+                isVerified = true;
+                break;
+            }
+
+            unchecked { ++index; }
+        }
+
+        if (!isVerified) {
+            revert InvalidSignature();
+        }   
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function addNode(
```

Comment:
Needs to support `establishmentDate` (see spec)

Reply:
Updated